### ISSUE TITLE
Led panel P5-6432-3528-16S-B4 with ICN2038s chip

### ISSUE THREAD

**Issue by maksun** [hzeller/rpi-rgb-led-matrix#1207]
Hello,

I just would like to know if someone here managed to make this panel working?
I'm using the active board from electrodragon, I'm trying to help a friend so I did some research and read the following:

"Check the controller chip that install the LED panel, if are ICN2038S or FM6126, those are not compatible Smartmatrix because these chips require three clock pulses by Latch."

I will tell my friend to try the led panel type FM6126 in case of... Any advice is welcome ;)

Thank you.

**Comment by maksun** [comment]
Hello,

Looks like this chipset is similar to the FM6126A as I managed to make it working with the "--led-panel-type=FM6126A" parameter. FYI, it was an ABCD panel style.

Regards,

**Comment by alsaleem00** [comment]
Hello I am able to display something on matrix. but refresh is bad.
If anyone is interested I can post more here

**Comment by wrldtvlr** [comment]
I have a P2 panel with this same chip. While I can display something on the panel, the addressing isn't right I would be interested in knowing more about what you've been able to do.

**Comment by alsaleem00** [comment]
most P2's have different different row driver (320x160mm has 1/40 scan). This requires handling the row driver in addition to ICN2038s. please check your matrix. 
If your matrix is 256x128mm then you only need the ICN2038s driver (scan will be 1/32 which is OK.)

**Comment by wrldtvlr** [comment]
Yes, physical size is 256x128, with 128x64 pixels and 1/32 scan.

**Comment by alsaleem00** [comment]
here is the code 
#define BITS_ON	MTX_RGB_PINS
#define BITS_OFF	0
#define ICN1038S_REG_1		11
#define ICN1038S_REG_2		12
#define ICN1038S_DEF_REG1	(0xf000)  // register 1. (11)
#define ICN1038S_DEF_REG2 	(0xf000)  // register 2. (12)
void matrix_write(uint16_t uVal, int width, int reg)
{
	uint16_t value;

	MTX_LAT_LOW();
	MTX_CLK_LOW();
	for (int i = 0; i < width; ++i) {
		value = (uVal & (0x8000 >> (i%16))) ? BITS_ON : BITS_OFF;
		if (i > (width - reg)) {
			MTX_LAT_HIGH();
		}
		PIN_LOW(MTX_RGB_PORT, BITS_ON);
		PIN_HIGH(MTX_RGB_PORT, value);
		MTX_CLK_LOW();
		MTX_CLK_LOW();
		MTX_CLK_LOW();
		MTX_CLK_HIGH();
		MTX_CLK_HIGH();
		MTX_CLK_HIGH();
	}
	MTX_CLK_LOW();
	MTX_CLK_LOW();
	MTX_LAT_LOW();
}
void icn1038s_init()
{
	matrix_write(ICN1038S_DEF_REG1, LED_WIDTH, ICN1038S_REG_1);
	matrix_write(ICN1038S_DEF_REG2, LED_WIDTH, ICN1038S_REG_2);
}
I repeated the clock/latch signals because my MCU is fast.

**Comment by jeroenverspeek** [comment]
@alsaleem00 looks like the right solution for the ICN2038S. I am trying to get a 128x128 panel to work with this chip (https://github.com/hzeller/rpi-rgb-led-matrix/issues/1713). 

Your code does not look like it is written for hzellers library. If you could give more (code) context that would be helpful.

**Comment by alsaleem00** [comment]
This code was written for STM32F722. 
The PIN_/MTX_CLK_/MTX_LAT_/MTX_CLK_to control the pin directly.

**Comment by jeroenverspeek** [comment]
Hi Alsaleem00,
Thanks for your reply.

Op za 19 okt. 2024 13:14 schreef alsaleem00 ***@***.***>:

**Comment by marcmerlin** [comment]
With apologies to everyone, this project is doing a one time cleanup of 
all the bugs that have not been addressed and closed in the last 10 years.

If you filed an issue to ask for help, please do not re-open this one or a new
bug, but go here instead: https://rpi-rgb-led-matrix.discourse.group/

Please be mindful that this project does not have any full time maintainers
or anyone paid. You should not expect personal help for your own projects or
answers to programming questions but you can ask for help with panels that
aren't working, setup questions, or anything that may be a bug here:
https://rpi-rgb-led-matrix.discourse.group/
If you need programming help and are willing to pay someone for it, please
state so in your message, and someone may be interested.

If you did report an actual bug that is still not fixed today, please accept
our apologies for this being mass closed, you can absolutely send a followup
and re-open the issue, or open a new up to date one as applicable.
Note that we very much welcome help, so if you can help fix your bug, you will
be a superstar :)

If your bug was linked to a pull request that was not resolved, please update
the pull request and the issue, and we will do our best to review them. We do
thank you for your undertanding that we have very limited free time, so any
extra work you do along with your bug report or PR request will be very much
appreciated and improve the odds that they will be looked at.

If you want to help contribute as a python or C# or any other language bindings,
you are also a superstar, please state so in the issue and we'll do our best to
give you the access and reviews you need.

For anything related to PWM panel support, please make sure you read README and
then refer to https://github.com/hzeller/rpi-rgb-led-matrix/issues/466 . 
Please do not comment if you're haven't read everything and/or if you're just
asking when your panel is going to be supported. Those bugs are if you're 
able to contribute to the effort.

For rPI5: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1603

For anything python, please link your bug to: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1828
For anything C#, please link your bug to: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1828
